libyul/optimiser/CircularReferencesPruner.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/CircularReferencesPruner.h>
+
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/Algorithms.h>
+
+using namespace std;
+using namespace solidity::yul;
+
+void CircularReferencesPruner::run(OptimiserStepContext& _context, Block& _ast)
+{
+	CircularReferencesPruner{_context.reservedIdentifiers}(_ast);
+}
+
+void CircularReferencesPruner::operator()(Block& _block)
+{
+	ASTModifier::operator()(_block);
```

Comment:
Maybe it is enough to assume that the function hoister has been called before this step, i.e. all function definitions are in the outermost block. In this case, the call to `ASTModifier::operator()(_block)` would not be needed and thus this would not traverse the whole ast - currently, `functionsToKeep` is re-generated for every single block in the ast.

Reply:
Done here 1893ac14af0919c91931b64ff2e352d25c6f342f